11_structures/main.go

```diff
@@ -157,25 +157,22 @@ func (t triangle) sumLengthSmallerLongestSide() bool {
 // - 1 if the area of the first triangle is greater than the area of the second
 // Hint: func (t1 triangle) compare(t2 triangle) int { ... }
 
-func semiParameter(side1, side2, side3 float64) float64 {
-	return (side1 + side2 + side3) / 0.5
+func (t triangle) semiParameter() float64 {
+	return (t.side1 + t.side2 + t.side3) / 0.5
 }
 
 // calculate square root
 // func sqrt(x float64) float64 {
 // }
 // squareRoot :=  math.Sqrt()
 
-func calcTriangleArea(semiParameter, side1, side2, side3 float64) float64 {
-	return math.Sqrt(semiParameter * (semiParameter * side1) * (semiParameter * side2) * (semiParameter * side3))
+func (t triangle) calcTriangleArea() float64 {
+	return math.Sqrt(t.semiParameter() * (t.semiParameter() * t.side1) * (t.semiParameter() * t.side2) * (t.semiParameter() * t.side3))
 }
 
 func (t1 triangle) compare(t2 triangle) int {
-	semiParametert1 := semiParameter(t1.side1, t1.side2, t1.side3)
-	calcTriangleAreat1 := calcTriangleArea(semiParametert1, t1.side1, t1.side2, t1.side3)
-
-	semiParametert2 := semiParameter(t2.side1, t2.side2, t2.side3)
-	calcTriangleAreat2 := calcTriangleArea(semiParametert2, t2.side1, t2.side2, t2.side3)
+	calcTriangleAreat1 := t1.calcTriangleArea()
+	calcTriangleAreat2 := t2.calcTriangleArea()
 	if calcTriangleAreat1 < calcTriangleAreat2 {
 		return -1
 	}
```

Comment:
@nobe4 What's the best practice, declare initial variables instead of repeatedly calling the functions?
I find both to be quite readable, depending on the variable/function naming.

Reply:
I'm not sure I understand, can you add examples of both ways in this thread?
FWIW, what is currently written in the file is fine :)

Reply:
`func (t triangle) calcTriangleArea() float64 {
	return math.Sqrt(t.semiPerimeter() * (t.semiPerimeter() * t.side1) * (t.semiPerimeter() * t.side2) * (t.semiPerimeter() * t.side3))
}

func (t1 triangle) compare(t2 triangle) int {
	calcTriangleAreat1 := t1.calcTriangleArea()
	calcTriangleAreat2 := t2.calcTriangleArea()
	if t1.calcTriangleArea() < t2.calcTriangleArea() {
		return -1
	}
	if t1.calcTriangleArea() == t2.calcTriangleArea() {
		return 0
	}
	if t1.calcTriangleArea() > t2.calcTriangleArea() {
		return 1
	}
	// debug, to remove
	return 6060
}
vs.

`func (t triangle) calcTriangleArea() float64 {
semiParameter := t.semiPerimeter()
	return math.Sqrt(semiParameter * semiParameter* t.side1) * (semiParameter * t.side2) * (semiParameter * t.side3))
}

func (t1 triangle) compare(t2 triangle) int {
	calcTriangleAreat1 := t1.calcTriangleArea()
	calcTriangleAreat2 := t2.calcTriangleArea()
	if calcTriangleAreat1 < calcTriangleAreat2 {
		return -1
	}
	if calcTriangleAreat1 == calcTriangleAreat2 {
		return 0
	}
	if calcTriangleAreat1 > calcTriangleAreat2 {
		return 1
	}
	// debug, to remove
	return 6060
}`

I did a mix of both.

Reply:
Ah I see! In most case, if you use the result of a function more than once, it's better to save it in a variable first. So your first example :)